### ISSUE TITLE
entities aren't created from groups with less than 2 members

### ISSUE THREAD

**Issue by ryanwinter** [zigpy/zha#47]
My situation here is that I have a group with 2 lights, and one of them failed so I ordered a new one from Ikea. However in the interim, because there was now only one light in the group, it was never discovered (from [this ](https://github.com/zigpy/zha/blob/41430355ecf993a2bfd82d5bd884416474d4fc68/zha/application/discovery.py#L562)logic) and therefor all my dashboards and automations that depended on that entity are broken.

My proposal would be to surface all groups, regardless of the number of members in situations like above.

**Comment by teskanoo** [comment]
I wholeheartedly concur. I experience the exact situation. I get it that mathematically one is not a group, but functionally, this logic breaks automations. 

I have a dead smart bulb to replace in a vaulted ceiling fixture, one of two bulbs in a group... Had to change my UI and automations to be able to turn it off manually or automatically.

**Comment by agoode** [comment]
I agree it would be useful to avoid this kind of filtering for exactly this reason: if a group of size 2 has a device failure, it shouldn't result in destroying the whole group.

**Comment by ryanwinter** [comment]
I'm wishing if this was implemented because so situations groups of 1 are automatically created. Perhaps a config option would work?

**Comment by agoode** [comment]
I've spent some time digging into the code. The assumption that groups have a minimum of 2 members is in a few places, so it's not super trivial to fix. There are tests that assume this as well, so it's taking a while to fix everything.

**Comment by puddly** [comment]
@agoode I would put your efforts on hold for a bit. This repository is not the ZHA you use in Home Assistant (yet) and not in a state to accept contributions and new features.

**Comment by agoode** [comment]
Ok, thanks! I did notice both repositories and assumed this one would eventually replace the one in HA.

**Comment by dmulcahey** [comment]
OK... started thinking about this. Here is a challenge: you add a fan, a light, and a switch to a group. What platform should the group entity get created for? One of the reasons for the 2 device with entities from the same domain restriction is so we know what platform to create the entity for.  

There are other edge cases as well... what happens when you have 1 light and 1 switch in a group? we could arbitrarily pick a platform to create the entity for... but then if you add another device for either platform we may swap the group entity platform and therefore the entity id... that would break stuff.  

Let me know if I am not explaining this clearly and I'll try to table out some examples.

**Comment by DarthSonic** [comment]
> OK... started thinking about this. Here is a challenge: you add a fan, a light, and a switch to a group. What platform should the group entity get created for? One of the reasons for the 2 device with entities from the same domain restriction is so we know what platform to create the entity for.
> 
> There are other edge cases as well... what happens when you have 1 light and 1 switch in a group? we could arbitrarily pick a platform to create the entity for... but then if you add another device for either platform we may swap the group entity platform and therefore the entity id... that would break stuff.
> 
> Let me know if I am not explaining this clearly and I'll try to table out some examples.

? If there is only one device, it is clear which domain to choose. I do not see how this connects to the request to enable groups with one device member only.

And for your comment - I do not know if this is technically possible - on group creation one could select the domain the group is for and then can only add devices of that domain.

**Comment by valenterry** [comment]
> OK... started thinking about this. Here is a challenge: you add a fan, a light, and a switch to a group. What platform should the group entity get created for? One of the reasons for the 2 device with entities from the same domain restriction is so we know what platform to create the entity for.
> 
> There are other edge cases as well... what happens when you have 1 light and 1 switch in a group?

Wait a second. This ticket is about allowing groups with a *single* member. Groups with multiple members (like 1 light and 1 switch) are already possible. The problem that are describing must already exist no?

> Let me know if I am not explaining this clearly and I'll try to table out some examples.

Yes, maybe I (or we) are missing something, but introducing groups with a single member should actually make things easier. In fact, I would argue that groups with zero members should be possible as well. For example, I'd like to setup my scripts and automations already without waiting for my lights to arrive. Then once they have arrived, I pair them and add them to the group and I'm already done.

**Comment by dmulcahey** [comment]
> > OK... started thinking about this. Here is a challenge: you add a fan, a light, and a switch to a group. What platform should the group entity get created for? One of the reasons for the 2 device with entities from the same domain restriction is so we know what platform to create the entity for.
> > There are other edge cases as well... what happens when you have 1 light and 1 switch in a group?
> 
> Wait a second. This ticket is about allowing groups with a _single_ member. Groups with multiple members (like 1 light and 1 switch) are already possible. The problem that are describing must already exist no?
> 
> > Let me know if I am not explaining this clearly and I'll try to table out some examples.
> 
> Yes, maybe I (or we) are missing something, but introducing groups with a single member should actually make things easier. In fact, I would argue that groups with zero members should be possible as well. For example, I'd like to setup my scripts and automations already without waiting for my lights to arrive. Then once they have arrived, I pair them and add them to the group and I'm already done.

Groups have 0 restrictions and can be used freely. These comments are about group entities. This is another reason I am hesitant to change things. There are 0 restrictions on groups themselves and they can be used in automations already in the way that folks are describing.

**Comment by valenterry** [comment]
> > > OK... started thinking about this. Here is a challenge: you add a fan, a light, and a switch to a group. What platform should the group entity get created for? One of the reasons for the 2 device with entities from the same domain restriction is so we know what platform to create the entity for.
> > > There are other edge cases as well... what happens when you have 1 light and 1 switch in a group?
> > 
> > 
> > Wait a second. This ticket is about allowing groups with a _single_ member. Groups with multiple members (like 1 light and 1 switch) are already possible. The problem that are describing must already exist no?
> > > Let me know if I am not explaining this clearly and I'll try to table out some examples.
> > 
> > 
> > Yes, maybe I (or we) are missing something, but introducing groups with a single member should actually make things easier. In fact, I would argue that groups with zero members should be possible as well. For example, I'd like to setup my scripts and automations already without waiting for my lights to arrive. Then once they have arrived, I pair them and add them to the group and I'm already done.
> 
> Groups have 0 restrictions and can be used freely. These comments are about group entities. This is another reason I am hesitant to change things. There are 0 restrictions on groups themselves and they can be used in automations already in the way that folks are describing.

The problem is that there are e.g. blueprints that only support entities. Or sometimes you have to then mix entities and single devices, which is also a bit annoying.

**Comment by tremby** [comment]
I just started using ZHA and was surprised to find single-member groups aren't fully supported.

My use case is that I have some groups which I expect to add more members to in future. If I can address automations etc to a group now, I won't have to go back and change them later when adding more devices.
